File: research-colossus/services/geminiService.ts
import { GoogleGenAI, Type } from "@google/genai";
import { AgentType, ResearchTask, SimulationDataPoint, TaskCritique, ResearchBranch, AgentExecutionResult, KnowledgeFact, ExperimentSpec } from "../types";

// Constants
const MODEL_REASONING = "gemini-3-pro-preview";
const MODEL_FAST = "gemini-2.5-flash";

// Initialize API
const ai = new GoogleGenAI({ apiKey: process.env.API_KEY });

/**
 * Step 1: Initial Planning with Isomorphic Reasoning
 */
export const orchestratePlan = async (prompt: string): Promise<ResearchBranch[]> => {
  try {
    const response = await ai.models.generateContent({
      model: MODEL_REASONING,
      contents: `You are the Research Colossus Orchestrator.
      
      USER QUERY: "${prompt}"
      
      STRATEGIC DIRECTIVE:
      Look for "Isomorphic Mappings". Example: Can concepts from Category Theory be mapped to this problem?
      
      Create a "Tree of Thoughts" research plan.
      For EACH branch, define the sequential tasks.
      
      CRITICAL: Assign a "priority" (1-10). 10 is immediate/critical.
      Tasks normally follow a logical flow (Hypothesis -> Math -> Simulation -> Conclusion).
      
      Return JSON structure.`,
      config: {
        responseMimeType: "application/json",
        responseSchema: {
          type: Type.ARRAY,
          items: {
            type: Type.OBJECT,
            properties: {
                id: { type: Type.STRING },
                name: { type: Type.STRING, description: "Name of this research branch" },
                tasks: {
                    type: Type.ARRAY,
                    items: {
                        type: Type.OBJECT,
                        properties: {
                            id: { type: Type.STRING },
                            description: { type: Type.STRING },
                            assignedTo: { type: Type.STRING },
                            priority: { type: Type.INTEGER, description: "1-10 Priority" },
                            status: { type: Type.STRING, enum: ["pending"] }
                        },
                        required: ["id", "description", "assignedTo", "status", "priority"]
                    }
                }
            },
            required: ["id", "name", "tasks"]
          },
        },
        thinkingConfig: { thinkingBudget: 4096 } // Increased for deep strategic planning
      },
    });

    const text = response.text;
    if (!text) throw new Error("No plan generated");
    const branches = JSON.parse(text) as ResearchBranch[];
    
    // Post-process to add sequential dependencies automatically for the initial plan
    return branches.map(b => {
        let prevTaskId: string | null = null;
        return {
            ...b,
            status: 'active',
            tasks: b.tasks.map((t, idx) => {
                const taskWithDeps: ResearchTask = {
                    ...t as any, // Cast to avoid strict type issues during mapping
                    iterationCount: 0,
                    critiques: [],
                    dependencies: prevTaskId ? [prevTaskId] : [],
                    status: 'pending'
                };
                prevTaskId = t.id;
                return taskWithDeps;
            })
        };
    });
  } catch (error) {
    console.error("Orchestration failed:", error);
    throw error;
  }
};

/**
 * Step 2: Task Execution with Knowledge Bank Context
 */
export const executeAgentTask = async (task: ResearchTask, knowledgeBank: KnowledgeFact[], feedback?: string): Promise<AgentExecutionResult> => {
  
  // Compile knowledge for context
  const knowledgeContext = knowledgeBank.map(k => `- [${k.sourceAgent}]: ${k.content}`).join('\n');

  let systemInstruction = `You are a world-class ${task.assignedTo}. 
  Your task is to perform deep research on: "${task.description}".
  
  KNOWN FACTS (Use these as axioms, do not reinvent):
  ${knowledgeContext || "No prior axioms."}`;

  if (feedback) {
    systemInstruction += `\n\nCRITICAL INSTRUCTION: Your previous attempt was critiqued. 
    Refine your work based on this feedback: "${feedback}".`;
  }
  
  systemInstruction += `\n\nProvide a detailed, scientific, and rigorous output. 
  - Use LaTeX for math formulas.
  - If a concept is abstract, try to ground it in a concrete analogy (Isomorphism).
  - If you use the search tool, cite the URLs provided.`;

  try {
    const response = await ai.models.generateContent({
      model: MODEL_REASONING,
      contents: task.description,
      config: {
        systemInstruction: systemInstruction,
        thinkingConfig: { thinkingBudget: 4096 },
        tools: [{ googleSearch: {} }] 
      },
    });

    const urls: string[] = [];
    if (response.candidates?.[0]?.groundingMetadata?.groundingChunks) {
        response.candidates[0].groundingMetadata.groundingChunks.forEach(chunk => {
            if (chunk.web?.uri) {
                urls.push(chunk.web.uri);
            }
        });
    }

    return {
        content: response.text || "Analysis complete (No output generated).",
        urls: [...new Set(urls)]
    };

  } catch (error) {
    console.error(`Agent ${task.assignedTo} failed:`, error);
    return { content: `Error executing task: ${error}`, urls: [] };
  }
};

/**
 * Step 3: Critique
 */
export const reviewTaskOutput = async (task: ResearchTask, output: string): Promise<TaskCritique> => {
    try {
        const response = await ai.models.generateContent({
            model: MODEL_REASONING,
            contents: `You are the "Peer Reviewer". Review this output from a ${task.assignedTo}.
            
            Task: "${task.description}"
            Output: "${output}"
            
            Standards:
            1. Is the math rigorous?
            2. Is the reasoning logically sound?
            3. If it's a specification, is it technically feasible?
            
            Rate 0-100. Threshold 85. Return JSON.`,
            config: {
                responseMimeType: "application/json",
                responseSchema: {
                    type: Type.OBJECT,
                    properties: {
                        score: { type: Type.INTEGER },
                        feedback: { type: Type.STRING },
                        approved: { type: Type.BOOLEAN }
                    },
                    required: ["score", "feedback", "approved"]
                },
                thinkingConfig: { thinkingBudget: 1024 }
            }
        });
        
        const text = response.text;
        if(!text) throw new Error("Review failed");
        return JSON.parse(text) as TaskCritique;
    } catch (error) {
        return { score: 100, feedback: "Review bypass.", approved: true };
    }
}

/**
 * Step 4: Manager Loop (Dynamic Re-planning & Triage)
 */
export const evaluateProgress = async (
    originalGoal: string, 
    lastTask: ResearchTask, 
    lastOutput: string, 
    knowledgeBank: KnowledgeFact[],
    allTasks: ResearchTask[] // To allow reprioritization
): Promise<{ newTasks: ResearchTask[], newFacts: KnowledgeFact[], stopBranch: boolean, reprioritize: {taskId: string, newPriority: number}[] }> => {
    
    try {
        // Simplified view of existing tasks for the context
        const taskState = allTasks.filter(t => t.status === 'pending').map(t => `ID: ${t.id} | Desc: ${t.description} | Pri: ${t.priority}`).join('\n');

        const response = await ai.models.generateContent({
            model: MODEL_REASONING,
            contents: `You are the Research Manager.
            Goal: "${originalGoal}"
            
            Just Completed: "${lastTask.description}" (Agent: ${lastTask.assignedTo}).
            Result Summary: "${lastOutput.substring(0, 1000)}..."
            
            Pending Tasks Context:
            ${taskState}
            
            1. Extract key FACTS.
            2. Determine if we need NEW tasks.
            3. TRIAGE: Do we need to change the priority of existing pending tasks based on this result? (e.g. if we found a blocker, move dependent tasks down, or if we found a lead, move specific tasks up).
            
            Return JSON.`,
            config: {
                responseMimeType: "application/json",
                responseSchema: {
                    type: Type.OBJECT,
                    properties: {
                        stopBranch: { type: Type.BOOLEAN },
                        newFacts: {
                            type: Type.ARRAY,
                            items: {
                                type: Type.OBJECT,
                                properties: {
                                    content: { type: Type.STRING },
                                    confidence: { type: Type.INTEGER }
                                }
                            }
                        },
                        newTasks: {
                            type: Type.ARRAY,
                            items: {
                                type: Type.OBJECT,
                                properties: {
                                    description: { type: Type.STRING },
                                    assignedTo: { type: Type.STRING },
                                    priority: { type: Type.INTEGER, description: "1-10"}
                                }
                            }
                        },
                        reprioritize: {
                            type: Type.ARRAY,
                            items: {
                                type: Type.OBJECT,
                                properties: {
                                    taskId: { type: Type.STRING },
                                    newPriority: { type: Type.INTEGER }
                                }
                            }
                        }
                    }
                },
                thinkingConfig: { thinkingBudget: 2048 }
            }
        });

        const data = JSON.parse(response.text || "{}");
        
        const newFacts = (data.newFacts || []).map((f: any) => ({
            id: Math.random().toString(36).substr(2, 9),
            content: f.content,
            sourceAgent: lastTask.assignedTo,
            timestamp: Date.now(),
            confidence: f.confidence
        }));

        const newTasks = (data.newTasks || []).map((t: any) => ({
            id: Math.random().toString(36).substr(2, 9),
            description: t.description,
            assignedTo: t.assignedTo,
            priority: t.priority || 5,
            status: 'pending',
            iterationCount: 0,
            critiques: [],
            dependencies: [lastTask.id] // Default dependency on the task that just finished
        }));

        return {
            stopBranch: data.stopBranch || false,
            newFacts,
            newTasks,
            reprioritize: data.reprioritize || []
        };

    } catch (e) {
        console.error("Manager eval failed", e);
        return { stopBranch: false, newFacts: [], newTasks: [], reprioritize: [] };
    }
}

/**
 * Step 5: Simulation OR Architecture Spec
 */
export const generateSimulationCode = async (
    taskDescription: string, 
    knowledgeBank: KnowledgeFact[],
    previousAttempt?: string 
): Promise<{type: 'CODE' | 'SPEC', code?: string, scenarios?: string[], spec?: ExperimentSpec}> => {
  
  const context = knowledgeBank.map(k => k.content).join('\n');

  try {
    const response = await ai.models.generateContent({
      model: MODEL_REASONING,
      contents: `You are a Scientific Computation Expert.
      
      Task: "${taskDescription}"
      Context: ${context}
      Previous Attempt (if any): ${previousAttempt || "None"}
      
      DECISION POINT:
      1. Is this a "Toy Model" that can run in a browser (Python/Numpy, < 1GB RAM, < 10s runtime)?
         -> If YES: Generate Python code using 'numpy' only (no matplotlib). Create global \`simulation_data\`.
         -> Output Type: "CODE"
      
      2. Is this a "Heavy Experiment" (e.g., Training 7B LLM, Large Scale CFD, Molecular Dynamics)?
         -> If YES: Do NOT try to run it. Instead, generate a "Cluster Specification" (A blueprint for a supercomputer).
         -> Output Type: "SPEC"
      
      Return JSON.`,
      config: {
        responseMimeType: "application/json",
        responseSchema: {
            type: Type.OBJECT,
            properties: {
                type: { type: Type.STRING, enum: ["CODE", "SPEC"] },
                // For CODE
                code: { type: Type.STRING, description: "Python code for Pyodide (if type=CODE)" },
                scenarios: { type: Type.ARRAY, items: { type: Type.STRING } },
                // For SPEC
                spec: {
                    type: Type.OBJECT,
                    properties: {
                        title: { type: Type.STRING },
                        complexity: { type: Type.STRING, enum: ["HIGH", "EXTREME"] },
                        requirements: { type: Type.ARRAY, items: { type: Type.STRING } },
                        codeSnippet: { type: Type.STRING, description: "The PyTorch/CUDA code for the cluster" },
                        hypothesis: { type: Type.STRING },
                        expectedOutcome: { type: Type.STRING }
                    }
                }
            },
        },
        thinkingConfig: { thinkingBudget: 4096 }
      }
    });

    const text = response.text;
    if (!text) throw new Error("No code generated");
    return JSON.parse(text);
  } catch (error) {
    console.error("Code generation failed:", error);
    throw error;
  }
};

export const synthesizeResearch = async (originalPrompt: string, knowledgeBank: KnowledgeFact[]): Promise<string> => {
  try {
    const knowledge = knowledgeBank.map(k => `[${k.sourceAgent}]: ${k.content}`).join('\n\n');
    const response = await ai.models.generateContent({
      model: MODEL_REASONING,
      contents: `Synthesize a Final Scientific Report.
      Query: "${originalPrompt}"
      
      Validated Knowledge Bank:
      ${knowledge}
      
      Format:
      1. Executive Summary
      2. Methodology (Reasoning Traces)
      3. Key Findings (proven via Simulation or Axioms)
      4. Future Work (Computational Specs created)
      `,
    });
    return response.text || "Synthesis failed.";
  } catch (error) {
    return "Could not synthesize final report.";
  }
};

File: app/services/gemini_service.py
from __future__ import annotations

import json
import logging
import os
import base64
from typing import Any, Dict, List, Tuple

from google import genai

logger = logging.getLogger(__name__)

# --- MODEL CONFIGURATION ---
# Free tier: Use "gemini-2.5-flash-lite" (1000 RPD, 15 RPM, 1M TPM)
# Paid tier: Uncomment below to use "gemini-3-pro-preview" for advanced reasoning
# MODEL_REASONING = "gemini-3-pro-preview"
MODEL_REASONING = os.getenv("GEMINI_REASONING_MODEL", "gemini-2.0-flash-exp")

# Speech model: fast, high-quota variant
MODEL_SPEECH = os.getenv("GEMINI_SPEECH_MODEL", "gemini-1.5-flash-8b")


class QuotaExhaustedError(Exception):
    """Raised when Gemini API quota/rate limits are hit."""
    pass


def get_client() -> genai.Client:
    api_key = os.getenv("GEMINI_API_KEY")
    if not api_key:
        raise RuntimeError("GEMINI_API_KEY environment variable is not set")
    return genai.Client(api_key=api_key)


def _handle_gemini_error(e: Exception, operation: str) -> None:
    """
    Check if the error is a quota/rate limit issue and raise a user-friendly exception.
    """
    error_str = str(e).lower()
    if "429" in error_str or "resource_exhausted" in error_str or "quota" in error_str:
        logger.error(f"{operation} failed due to quota: {e}")
        raise QuotaExhaustedError(
            f"Gemini API quota exhausted during {operation}. "
            "Please wait for quota reset or upgrade to a paid plan. "
            "See: https://ai.google.dev/pricing"
        ) from e
    # Re-raise other errors as-is
    logger.error(f"{operation} failed: {e}")
    raise


async def orchestrate_plan(prompt: str, attachments: List[Dict[str, Any]] | None = None) -> List[Dict[str, Any]]:
    """
    Ask Gemini to create a research plan with branches and tasks.
    Returns a list of branches with embedded tasks.
    """
    client = get_client()

    attachment_text = ""
    if attachments:
        chunks: List[str] = []
        for att in attachments:
            name = att.get("name", "attachment")
            content = att.get("content", "")
            # Truncate very large files to keep prompt size reasonable
            if len(content) > 4000:
                content = content[:4000] + "\n...[truncated]..."
            chunks.append(f"FILE: {name}\n{content}")
        attachment_text = "\n\nATTACHED CONTEXT:\n" + "\n\n".join(chunks)

    system_prompt = (
        "You are the Research Colossus Orchestrator.\n"
        "Create a Tree-of-Thoughts style research plan.\n"
        "For each branch, define a name and a sequence of tasks.\n"
        "Each task should include: id, description, assigned_to (role), priority (1-10), status ('pending').\n"
        "Return strictly valid JSON matching this structure:\n"
        "[{\n"
        '  "id": "branch-1",\n'
        '  "name": "Branch Name",\n'
        '  "tasks": [\n'
        '    {"id": "task-1", "description": "...", "assigned_to": "Theoretical Physicist", "priority": 8, "status": "pending"}\n'
        "  ]\n"
        "}]\n"
    )

    user_text = f"{system_prompt}\n\nUSER PROMPT: {prompt}"
    if attachment_text:
        user_text += f"\n\n{attachment_text}"

    try:
        resp = client.models.generate_content(
            model=MODEL_REASONING,
            contents=[{"role": "user", "parts": [{"text": user_text}]}],
        )
        raw = resp.text or ""
        logger.debug("orchestrate_plan raw response: %s", raw)
    except Exception as e:
        _handle_gemini_error(e, "orchestrate_plan")

    # Be defensive about extra text / code fences
    try:
        data = _extract_json(raw)
    except json.JSONDecodeError as exc:
        logger.error("Failed to parse orchestrate_plan JSON: %s", exc)
        raise

    if not isinstance(data, list):
        raise ValueError("orchestrate_plan: expected a list of branches")
    return data


async def execute_agent_task(task_desc: str, role: str, context: str) -> Tuple[str, List[str]]:
    """
    Execute a single agent task with optional context.
    Returns (content, [urls]).
    """
    client = get_client()

    system_instruction = (
        f"You are a world-class {role}. "
        "Perform deep scientific reasoning on the given task.\n"
        "Use the provided context as axioms when helpful.\n"
        "Provide rigorous and detailed output."
    )

    contents = [
        {
            "role": "user",
            "parts": [
                {"text": system_instruction},
                {"text": f"TASK: {task_desc}"},
                {"text": f"CONTEXT:\n{context or 'No prior context.'}"},
            ],
        }
    ]

    try:
        resp = client.models.generate_content(
            model=MODEL_REASONING,
            contents=contents,
        )
        text = resp.text or ""
    except Exception as e:
        _handle_gemini_error(e, "execute_agent_task")

    urls: List[str] = []
    try:
        candidates = getattr(resp, "candidates", []) or []
        gm = getattr(candidates[0], "grounding_metadata", None)
        if gm and getattr(gm, "grounding_chunks", None):
            for chunk in gm.grounding_chunks:
                if getattr(chunk, "web", None) and getattr(chunk.web, "uri", None):
                    urls.append(chunk.web.uri)
    except Exception:  # pragma: no cover - best-effort extraction
        logger.debug("No grounding URLs found in response.")

    # Deduplicate
    urls = list(dict.fromkeys(urls))
    return text, urls


async def review_task_output(task: str, output: str) -> Dict[str, Any]:
    """
    Peer-review a task result. Returns a dict with score, feedback, approved.
    """
    client = get_client()

    prompt = (
        "You are the Peer Reviewer for Research Colossus.\n"
        "Review the output for rigor, coherence, and correctness.\n"
        "Rate from 0-100 and set approved=true if score >= 85.\n"
        "Return strictly JSON: {\"score\": int, \"feedback\": str, \"approved\": bool}.\n\n"
        f"TASK: {task}\n\nOUTPUT:\n{output}\n"
    )

    try:
        resp = client.models.generate_content(
            model=MODEL_REASONING,
            contents=[{"role": "user", "parts": [{"text": prompt}]}],
        )
        raw = resp.text or ""
        logger.debug("review_task_output raw response: %s", raw)
    except Exception as e:
        _handle_gemini_error(e, "review_task_output")

    try:
        data = _extract_json(raw)
    except json.JSONDecodeError:
        # Fallback: accept with generic feedback
        return {"score": 100, "feedback": "Automatic approval (parse failure).", "approved": True}

    if not isinstance(data, dict):
        raise ValueError("review_task_output: expected a JSON object")
    return data


async def generate_simulation_code(task: str, context: str) -> Dict[str, Any]:
    """
    Decide between generating toy-model Python (for browser / local sim) or an HPC spec.
    Returns a JSON-like dict describing either {type: 'CODE', code, scenarios}
    or {type: 'SPEC', spec: {...}}.
    """
    client = get_client()

    prompt = (
        "You are a Scientific Computation Expert for Research Colossus.\n"
        "Given a task and context, decide whether to create a toy-model Python simulation\n"
        "(numpy only, small runtime) or a heavy cluster specification.\n"
        "If toy model, return:\n"
        '{\"type\": \"CODE\", \"code\": \"python code here\", \"scenarios\": [\"label1\", \"label2\"]}\n'
        "If heavy experiment, return:\n"
        "{\n"
        '  \"type\": \"SPEC\",\n'
        '  \"spec\": {\n'
        '    \"title\": \"...\",\n'
        '    \"complexity\": \"HIGH\" or \"EXTREME\",\n'
        '    \"requirements\": [\"8x H100 GPUs\", \"PyTorch 2.1\"],\n'
        '    \"codeSnippet\": \"CUDA / PyTorch code\",\n'
        '    \"hypothesis\": \"...\",\n'
        '    \"expectedOutcome\": \"...\"\n'
        "  }\n"
        "}\n"
        "Return strictly valid JSON with one of these shapes.\n\n"
        f"TASK: {task}\n\nCONTEXT:\n{context}\n"
    )

    try:
        resp = client.models.generate_content(
            model=MODEL_REASONING,
            contents=[{"role": "user", "parts": [{"text": prompt}]}],
        )
        raw = resp.text or ""
    except Exception as e:
        _handle_gemini_error(e, "generate_simulation_code")
    logger.debug("generate_simulation_code raw response: %s", raw)

    data = _extract_json(raw)
    if not isinstance(data, dict):
        raise ValueError("generate_simulation_code: expected an object")
    return data


async def synthesize_research(original_prompt: str, knowledge_facts: List[Dict[str, Any]]) -> str:
    """
    Produce a final scientific report based on the validated knowledge bank.
    `knowledge_facts` is a list of dicts with at least `source_agent` and `content` keys.
    """
    client = get_client()

    knowledge_str = "\n\n".join(
        f"[{fact.get('source_agent', 'Agent')}]: {fact.get('content', '')}" for fact in knowledge_facts
    )

    prompt = (
        "Synthesize a Final Scientific Report.\n"
        f"Query: \"{original_prompt}\"\n\n"
        "Validated Knowledge Bank:\n"
        f"{knowledge_str}\n\n"
        "Format:\n"
        "1. Executive Summary\n"
        "2. Methodology (Reasoning Traces)\n"
        "3. Key Findings (proven via Simulation or Axioms)\n"
        "4. Future Work (Computational Specs created)\n"
    )

    try:
        resp = client.models.generate_content(
            model=MODEL_REASONING,
            contents=[{"role": "user", "parts": [{"text": prompt}]}],
        )
        return resp.text or "Synthesis failed."
    except Exception as exc:  # pragma: no cover - defensive
        logger.error("synthesize_research failed: %s", exc)
        return "Could not synthesize final report."


async def transcribe_audio(audio_bytes: bytes, mime_type: str) -> str:
    """
    Use Gemini to perform speech-to-text on an audio clip.
    `mime_type` should be a valid audio MIME type (e.g. audio/webm, audio/wav).
    """
    client = get_client()

    # Encode audio as base64 for inline_data
    encoded = base64.b64encode(audio_bytes).decode("utf-8")

    try:
        resp = client.models.generate_content(
            model=MODEL_SPEECH,
            contents=[
                {
                    "role": "user",
                    "parts": [
                        {"text": "Transcribe this audio into clear English text."},
                        {
                            "inline_data": {
                                "data": encoded,
                                "mime_type": mime_type or "audio/webm",
                            }
                        },
                    ],
                }
            ],
        )
        text = resp.text or ""
        return text.strip()
    except Exception as exc:  # pragma: no cover - defensive
        logger.error("transcribe_audio failed: %s", exc)
        raise


def _extract_json(raw: str) -> Any:
    """
    Best-effort extraction of JSON from model output, stripping code fences or prose.
    """
    text = raw.strip()

    # Strip common Markdown fences
    if text.startswith("```"):
        # Drop first line and last fence if present
        lines = text.splitlines()
        # remove leading and trailing ```...``` lines
        if lines and lines[0].startswith("```"):
            lines = lines[1:]
        if lines and lines[-1].startswith("```"):
            lines = lines[:-1]
        text = "\n".join(lines).strip()

    # Heuristic: find first '{' or '[' and last '}' or ']'
    start_candidates = [i for i in (text.find("{"), text.find("[")) if i != -1]
    if start_candidates:
        start = min(start_candidates)
        # naive end search
        end_brace = text.rfind("}")
        end_bracket = text.rfind("]")
        end = max(end_brace, end_bracket)
        if end != -1 and end >= start:
            text = text[start : end + 1]

    return json.loads(text)


